crates/amaru-ledger/src/state.rs

```diff
@@ -541,6 +541,26 @@ impl<S: Store, HS: HistoricalStores> State<S, HS> {
         Ok(result)
     }
 
+    /// Calcualte chain density over the last `k` blocks (or oldest block in the volatileDB) given some `Point`.
+    /// If the `Point` is older than the oldest block in the volatileDB, density is 0
+    pub fn chain_density(&self, point: &Point) -> f64 {
+        let latest_slot = point.slot_or_default();
+        let k_slot = self
+            .volatile
+            .view_front()
+            .map(|state| &state.anchor.0)
+            .unwrap_or(&Point::Origin)
+            .slot_or_default();
+
+        if k_slot > latest_slot {
+            0f64
+        } else {
+            // Add one to the number of blocks in the volatileDB because we are including the `Point` in the chain density
+            (self.volatile.len() as f64 + 1f64)
+                / (u64::from(latest_slot) as f64 - u64::from(k_slot) as f64)
```

Comment:
The dimensional analysis doesn't quite check for this one; this is a ratio of number of _blocks_ over a number of _slots_. I believe the numerator should be:

- k  

and the denominator:

- (latest_slot - k_slot) * active_slot_coeff_inverse 

So said differently, we compute the actual number of blocks created over the "maximum" that could have been created during that time. Ideally, we want the numerator to be as close to k * active_slot_coeff_inverse (43200 on mainnet) as possible.

No sure about the `+1` either.

Reply:
_sigh_, I guess that's one way to look at it. Though that means an expected "ideal" value for the chain density is ~5%, which just sounds extremely weird to me. 

I could trace back where this ... oddity comes from; and it seems that back in 2019, people just translated the Ouroboros Classic "formula" to Praos'. 

https://github.com/IntersectMBO/cardano-node/issues/250

This worked in Classic because you expect (in nominal conditions) one block per slot; and so an ideal density is equal to `1`. But with Praos and the notion of active slot, that doesn't hold anymore and it seems like no one bothered about doing a dimensional analysis. 

I'll likely open an issue on the cardano-node... though I don't know if I have the energy to fight for it.